// azurite/dist/src/table/persistence/QueryTranscriber/PredicateModel/DatePredicate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const TaggedToken_1 = tslib_1.__importDefault(require("../TokenModel/TaggedToken"));
const IdentifierToken_1 = tslib_1.__importDefault(require("../TokenModel/IdentifierToken"));
const ValueToken_1 = tslib_1.__importDefault(require("../TokenModel/ValueToken"));
class DatePredicate {
    constructor(tokenMap) {
        this.tokenMap = tokenMap;
    }
    /**
     * converts a datetime predicate for lokijs schema
     *
     * @return {*}
     * @memberof DatePredicate
     */
    convertPredicateForLokiJS() {
        const newTokens = [];
        this.tokenMap.tokens.forEach((taggedToken) => {
            this.pushValue(taggedToken, newTokens);
            this.pushIdentifier(taggedToken, newTokens);
            this.pushOperator(taggedToken, newTokens);
        });
        this.tokenMap.tokens = newTokens;
        return this;
    }
    /**
     * pushes value for date predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DatePredicate
     */
    pushValue(taggedToken, newTokens) {
        if (taggedToken.type.isValue()) {
            newTokens.push(new TaggedToken_1.default("new Date(" +
                taggedToken.token.substring(8, taggedToken.token.length) +
                ").getTime()", new ValueToken_1.default()));
        }
    }
    /**
     * pushes identifier in a date predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DatePredicate
     */
    pushIdentifier(taggedToken, newTokens) {
        if (taggedToken.type.isIdentifier()) {
            newTokens.push(new TaggedToken_1.default(`new Date(item.properties.${taggedToken.token}).getTime()`, new IdentifierToken_1.default()));
        }
    }
    /**
     * Pushes operator in a date predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DatePredicate
     */
    pushOperator(taggedToken, newTokens) {
        if (taggedToken.type.isOperator()) {
            newTokens.push(taggedToken);
        }
    }
}
exports.default = DatePredicate;
