// azurite/dist/src/table/persistence/QueryTranscriber/PredicateModel/DoublePredicate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const TaggedToken_1 = tslib_1.__importDefault(require("../TokenModel/TaggedToken"));
const IdentifierToken_1 = tslib_1.__importDefault(require("../TokenModel/IdentifierToken"));
const ValueToken_1 = tslib_1.__importDefault(require("../TokenModel/ValueToken"));
class DoublePredicate {
    constructor(tokenMap) {
        this.tokenMap = tokenMap;
    }
    /**
     * converts a double predicate for lokijs schema
     *
     * @return {*}
     * @memberof DoublePredicate
     */
    convertPredicateForLokiJS() {
        const newTokens = [];
        this.tokenMap.tokens.forEach((taggedToken) => {
            this.pushValue(taggedToken, newTokens);
            this.pushIdentifier(taggedToken, newTokens);
            this.pushOperator(taggedToken, newTokens);
        });
        this.tokenMap.tokens = newTokens;
        return this;
    }
    /**
     * pushes value for double predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DoublePredicate
     */
    pushValue(taggedToken, newTokens) {
        if (taggedToken.type.isValue()) {
            newTokens.push(new TaggedToken_1.default(taggedToken.token, new ValueToken_1.default()));
        }
    }
    /**
     * pushes identifier for double predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DoublePredicate
     */
    pushIdentifier(taggedToken, newTokens) {
        if (taggedToken.type.isIdentifier()) {
            newTokens.push(new TaggedToken_1.default(`item.properties.${taggedToken.token}`, new IdentifierToken_1.default()));
        }
    }
    /**
     * pushes operator for double predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof DoublePredicate
     */
    pushOperator(taggedToken, newTokens) {
        if (taggedToken.type.isOperator()) {
            newTokens.push(taggedToken);
        }
    }
}
exports.default = DoublePredicate;
