// azurite/dist/src/table/persistence/QueryTranscriber/PredicateModel/IntegerPredicate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const TaggedToken_1 = tslib_1.__importDefault(require("../TokenModel/TaggedToken"));
const IdentifierToken_1 = tslib_1.__importDefault(require("../TokenModel/IdentifierToken"));
const ValueToken_1 = tslib_1.__importDefault(require("../TokenModel/ValueToken"));
class IntegerPredicate {
    constructor(tokenMap) {
        this.tokenMap = tokenMap;
    }
    /**
     * Converts an integer predicate for lokijs schema
     *
     * @return {*}
     * @memberof IntegerPredicate
     */
    convertPredicateForLokiJS() {
        const newTokens = [];
        this.tokenMap.tokens.forEach((taggedToken) => {
            this.pushValue(taggedToken, newTokens);
            this.pushIdentifier(taggedToken, newTokens);
            this.pushOperator(taggedToken, newTokens);
        });
        this.tokenMap.tokens = newTokens;
        return this;
    }
    /**
     * Pushes the value to the integer predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof IntegerPredicate
     */
    pushValue(taggedToken, newTokens) {
        if (taggedToken.type.isValue()) {
            newTokens.push(new TaggedToken_1.default(taggedToken.token, new ValueToken_1.default()));
        }
    }
    /**
     * Pushes the identifier to the integer predicate
     *
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof IntegerPredicate
     */
    pushIdentifier(taggedToken, newTokens) {
        if (taggedToken.type.isIdentifier()) {
            newTokens.push(new TaggedToken_1.default(`item.properties.${taggedToken.token}`, new IdentifierToken_1.default()));
        }
    }
    /**
     * Pushes the operator to the integer predicate
     *
     * @private
     * @param {TaggedToken} taggedToken
     * @param {TaggedToken[]} newTokens
     * @memberof IntegerPredicate
     */
    pushOperator(taggedToken, newTokens) {
        if (taggedToken.type.isOperator()) {
            newTokens.push(taggedToken);
        }
    }
}
exports.default = IntegerPredicate;
