// azurite/dist/src/table/persistence/QueryTranscriber/QueryStateName.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryStateName = void 0;
/**
 * Query state names are used in state store to
 * access and identify states
 *
 * @export
 * @enum {number}
 */
var QueryStateName;
(function (QueryStateName) {
    QueryStateName[QueryStateName["None"] = 0] = "None";
    QueryStateName[QueryStateName["QueryStarted"] = 1] = "QueryStarted";
    QueryStateName[QueryStateName["QueryFinished"] = 2] = "QueryFinished";
    QueryStateName[QueryStateName["PredicateStarted"] = 3] = "PredicateStarted";
    QueryStateName[QueryStateName["PredicateFinished"] = 4] = "PredicateFinished";
    QueryStateName[QueryStateName["ProcessIdentifier"] = 5] = "ProcessIdentifier";
    QueryStateName[QueryStateName["ProcessOperator"] = 6] = "ProcessOperator";
    QueryStateName[QueryStateName["ProcessValue"] = 7] = "ProcessValue";
    QueryStateName[QueryStateName["ProcessPredicateOperator"] = 8] = "ProcessPredicateOperator";
    QueryStateName[QueryStateName["ProcessParensOpen"] = 9] = "ProcessParensOpen";
    QueryStateName[QueryStateName["ProcessParensClose"] = 10] = "ProcessParensClose";
})(QueryStateName = exports.QueryStateName || (exports.QueryStateName = {}));
