// azurite/dist/src/table/persistence/QueryTranscriber/StateProcessOperator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const QPState_1 = tslib_1.__importDefault(require("./QPState"));
const QueryStateName_1 = require("./QueryStateName");
const TaggedToken_1 = tslib_1.__importDefault(require("./TokenModel/TaggedToken"));
const OperatorToken_1 = tslib_1.__importDefault(require("./TokenModel/OperatorToken"));
/**
 * contains logic to handle operators
 *
 * @export
 * @class StateProcessOperator
 * @extends {QPState}
 * @implements {IQPState}
 */
class StateProcessOperator extends QPState_1.default {
    constructor() {
        super(...arguments);
        this.name = QueryStateName_1.QueryStateName.ProcessOperator;
        /**
         * process current query token which is operator
         *
         * @param {QueryContext} context
         * @memberof StateProcessOperator
         */
        this.onProcess = (context) => {
            let token = "";
            [context, token] = this.getNextToken(context);
            const originalTokenLength = token.length;
            token = this.convertOperatorToken(token);
            context = this.storeTaggedTokens(context, token, originalTokenLength);
            [context, token] = this.getNextToken(context);
            context = this.handleToken(context, token);
            return context;
        };
        /**
         * optional post processing, here we can add logging
         * or additional validation etc
         *
         * @param {QueryContext} context
         * @memberof StateProcessValue
         */
        this.onExit = (context) => {
            return context;
        };
    }
    /**
     * state transition logic
     *
     * @protected
     * @param {QueryContext} context
     * @param {string} token
     * @return {*}  {QueryContext}
     * @memberof StateProcessOperator
     */
    handleToken(context, token) {
        // categorize the token
        if (token === "") {
            context.stateQueue.push(QueryStateName_1.QueryStateName.PredicateFinished);
        }
        else if (token === "(") {
            context.stateQueue.push(QueryStateName_1.QueryStateName.PredicateStarted);
        }
        else if (token === ")") {
            context.stateQueue.push(QueryStateName_1.QueryStateName.PredicateFinished);
        }
        else if (this.isPredicateOperator(token)) {
            if (this.name === QueryStateName_1.QueryStateName.PredicateFinished) {
                context.stateQueue.push(QueryStateName_1.QueryStateName.ProcessPredicateOperator);
            }
            else {
                context.stateQueue.push(QueryStateName_1.QueryStateName.PredicateFinished);
            }
            // will need to end current predicate and create a new predicate
        }
        else if (this.isOperand(token)) {
            // match operand (specific set)
            context.stateQueue.push(QueryStateName_1.QueryStateName.ProcessOperator);
        }
        else if (this.isValue(token)) {
            // match number (long & doubles? needed)
            // match string (starts with ', or " ?)
            // match guid (is exactly guid'<guidval>')
            context.stateQueue.push(QueryStateName_1.QueryStateName.ProcessValue);
        }
        else if (this.isIdentifier(token)) {
            // match identifier (can only start with letter)
            context.stateQueue.push(QueryStateName_1.QueryStateName.ProcessIdentifier);
        }
        return context;
    }
    /**
     * Converts a token from query request to a type used in persistence
     * layer query function.
     *
     * @private
     * @static
     * @param {string} token
     * @return {*}  {string}
     * @memberof LokiTableMetadataStore
     */
    convertOperatorToken(token) {
        switch (token) {
            case "eq":
                return "===";
            case "gt":
                return ">";
            case "ge":
                return ">=";
            case "lt":
                return "<";
            case "le":
                return "<=";
            case "ne":
                return "!==";
            case "and":
                return "&&";
            case "or":
                return "||";
            case "not":
                return "!";
            default:
                return token;
        }
    }
    /**
     * stores and tags the token as an operator
     *
     * @private
     * @param {QueryContext} context
     * @param {string} token
     * @param {number} originalTokenLength
     * @return {*}  {QueryContext}
     * @memberof StateProcessOperator
     */
    storeTaggedTokens(context, token, originalTokenLength) {
        const taggedToken = new TaggedToken_1.default(token, new OperatorToken_1.default());
        const taggedTokens = this.updateTaggedTokens(context, taggedToken);
        context = this.updateTaggedPredicate(taggedTokens, context);
        context.currentPos += originalTokenLength;
        return context;
    }
}
exports.default = StateProcessOperator;
