// azurite/dist/src/table/persistence/QueryTranscriber/StateProcessParensClose.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const QPState_1 = tslib_1.__importDefault(require("./QPState"));
const QueryStateName_1 = require("./QueryStateName");
const TaggedToken_1 = tslib_1.__importDefault(require("./TokenModel/TaggedToken"));
const TokenMap_1 = require("./PredicateModel/TokenMap");
const ParensCloseToken_1 = tslib_1.__importDefault(require("./TokenModel/ParensCloseToken"));
const ParensClose_1 = tslib_1.__importDefault(require("./PredicateModel/ParensClose"));
/**
 * contains the logic to handle parens close
 *
 * @export
 * @class StateProcessParensClose
 * @extends {QPState}
 * @implements {IQPState}
 */
class StateProcessParensClose extends QPState_1.default {
    constructor() {
        super(...arguments);
        this.name = QueryStateName_1.QueryStateName.ProcessParensClose;
        /**
         * process current query token which is operator
         *
         * @param {QueryContext} context
         * @memberof StateProcessOperator
         */
        this.onProcess = (context) => {
            let token = "";
            [context, token] = this.getNextToken(context);
            context = this.storeTaggedTokens(context, token);
            [context, token] = this.getNextToken(context);
            context = this.handleToken(context, token);
            return context;
        };
        /**
         * optional post processing, here we can add logging
         * or additional validation etc
         *
         * @param {QueryContext} context
         * @memberof StateProcessValue
         */
        this.onExit = (context) => {
            return context;
        };
    }
    /**
     * state transition logic
     *
     * @protected
     * @param {QueryContext} context
     * @param {string} token
     * @return {*}  {QueryContext}
     * @memberof StateProcessParensClose
     */
    handleToken(context, token) {
        // Parens Close will always end the predicate
        context.stateQueue.push(QueryStateName_1.QueryStateName.PredicateFinished);
        return context;
    }
    /**
     * Stores the tokens
     *
     * @private
     * @param {QueryContext} context
     * @param {string} token
     * @param {number} originalTokenLength
     * @return {*}  {QueryContext}
     * @memberof StateProcessParensClose
     */
    storeTaggedTokens(context, token) {
        const taggedParensToken = new TaggedToken_1.default(token, new ParensCloseToken_1.default());
        const parensTokenMap = new TokenMap_1.TokenMap([taggedParensToken]);
        context.taggedPredicates.push(new ParensClose_1.default(parensTokenMap));
        context.currentPredicate += 1;
        context.currentPos += token.length;
        return context;
    }
}
exports.default = StateProcessParensClose;
